<a href="https://colab.research.google.com/github/Dmitriy6655/-python-DS-ITOGOVAYA/blob/main/sem_10_itog_HW.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Урок 10. Практическое применение метода K-Means. Иерархическая кластеризация
Цель: Проанализировать данные о заработной плате сотрудников из файла ds_salaries.csv и построить модели машинного обучения для прогнозирования заработной платы.

Задачи:

1. Первичный анализ данных:
— Импортировать данные из файла ds_salaries.csv.
— Очистить данные от дубликатов и пропущенных значений.
— Исследовать типы данных, описательную статистику и корреляции между признаками.
— Визуализировать данные с помощью Matplotlib и Seaborn.

2. Построение модели линейной регрессии:
— Разделить данные на обучающую и тестовую выборки.
— Построить модель линейной регрессии для прогнозирования заработной платы.
— Оценить качество модели с помощью метрик MAE, MSE и R-квадрат.
— Проанализировать значимость признаков и мультиколлинеарность.

3. Построение модели дерева решений:
— Построить модель дерева решений для прогнозирования заработной платы.
— Оценить качество модели с помощью метрик RMSE и R-квадрат.
— Подберите гиперпараметры для модели дерево решений
— Визуализировать дерево решений.

4. Построение ансамблевой модели:
— Построить ансамблевую модель, используя Random Forest и Boosting.
— Сравнить качество ансамблевой модели с отдельными моделями.
— Оценить важность признаков в ансамблевой модели.

5. Понижение размерности:
— Применить Principal Component Analysis (PCA) для снижения размерности данных.
— Построить модель линейной регрессии с использованием данных, полученных после PCA.
— Сравнить качество модели с использованием исходных данных.

7. Кластеризация:
— Применить метод K-Means для кластеризации данных по заработной плате.
— Определить оптимальное количество кластеров.
— Проанализировать характеристики полученных кластеров.
— Визуализировать результаты кластеризации и проанализировать характеристики в каждом кластере.
— Проанализируйте полученные данные и сделайте выводы о том, какой алгоритм кластеризации сможет выделить кластеры лучше, чем K-Means.

Информация о наборе данных:



*   work_year - рабочий год;
*   experience_level - уровень опыта;
* employment type - тип занятости;
* job_title -должность;
* salary - уровень ЗП;
* salary_currency - валюта ЗП;
* salary_in_usd - ЗП в USD;
* employee_residence - служебное жилье;
* remote_ratio - удаленная работа;
* company_location - месторасположение компании;
* company_size - индекс количества сотрудников в комапнии.











Импортировать данные из файла ds_salaries.csv.

In [3]:
import warnings
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.colors import ListedColormap

###Первичный анализ

Посмотрим первые 5 строк данных:

In [12]:
ds_salaries = pd.read_csv('ds_salaries.csv')
ds_salaries.head()


work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

Посмотрим последние 5 строк данных:

In [11]:
ds_salaries.tail()

work_year experience_level employment_type                 job_title  \
3750       2020               SE              FT            Data Scientist   
3751       2021               MI              FT  Principal Data Scientist   
3752       2020               EN              FT            Data Scientist   
3753       2020               EN              CT     Business Data Analyst   
3754       2021               SE              FT      Data Science Manager   

       salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
3750   412000             USD         412000                 US           100   
3751   151000             USD         151000                 US           100   
3752   105000             USD         105000                 US           100   
3753   100000             USD         100000                 US           100   
3754  7000000             INR          94665                 IN            50   

     company_location company_size  
3750               US            L  
3751               US            L  
3752               US            S  
3753               US            L  
3754               IN            L

In [7]:
ds_salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


Убедимся что нет пропусков данных:

In [13]:
ds_salaries.isna().sum()

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64

1. Данны состоят из 3755 наблюдений
(строк) и 11 переменных (столбцов).
2. Премееные состоят из следюущих типов: 4 целочисленный тип данных, 7 объектов.
3. Нет пропущенных значений в данных, так как для каждой переменной количество non-null значений равно общему количеству наблюдений (3755).
4. Переменная 'salary' является целевой переменной (зависимой переменной), которую нужно предсказывать или анализировать в зависимости от других переменных.
5. Переменные 'work_year', 'salary_in_usd', 'remote_ratio' представляют собой категориальные переменные, так как имеют целочисленный тип данных.


###Статистические характеристики признаков

Для анализа и интерпретации необходимо преобразовать 9 столбцов (cp, fbs, restecg, slope, sex, ca, thal, exang и target) из числовых в строковый тип данных, несмотря на их числовую природу. Такой вывод основан на рассмотрении типов данных и объяснений функций, представленных ранее.

In [8]:
# Определим непрерывные признаки
continuous_features = ['work_year', 'salary', 'salary_in_usd', 'remote_ratio']


# Определим признаки, которые будут преобразованы в объектный тип данных
features_to_convert = [feature for feature in ds_salaries.columns if feature not in continuous_features]

# Преобразуем идентифицированные объекты в объектный тип данных
ds_salaries[features_to_convert] = ds_salaries[features_to_convert].astype('object')

ds_salaries.dtypes

work_year              int64
experience_level      object
employment_type       object
job_title             object
salary                 int64
salary_currency       object
salary_in_usd          int64
employee_residence    object
remote_ratio           int64
company_location      object
company_size          object
dtype: object

 df.describe().T - функция транспонирования сводной статистики в библиотеке Pandas.

Она позволяет поменять оси местами, и по горизонтали будут уже статистические параметры.

In [9]:
ds_salaries.describe().T

count           mean            std     min       25%  \
work_year      3755.0    2022.373635       0.691448  2020.0    2022.0   
salary         3755.0  190695.571771  671676.500508  6000.0  100000.0   
salary_in_usd  3755.0  137570.389880   63055.625278  5132.0   95000.0   
remote_ratio   3755.0      46.271638      48.589050     0.0       0.0   

                    50%       75%         max  
work_year        2022.0    2023.0      2023.0  
salary         138000.0  180000.0  30400000.0  
salary_in_usd  135000.0  175000.0    450000.0  
remote_ratio        0.0     100.0       100.0

Выясним количество сотрудников которые работаают и удаленно и в офисе:

In [15]:
ds_salaries.remote_ratio.value_counts()

remote_ratio
0      1923
100    1643
50      189
Name: count, dtype: int64

In [16]:
ds_salaries.work_year.value_counts()

work_year
2023    1785
2022    1664
2021     230
2020      76
Name: count, dtype: int64

На основе представленных статистических данных можно сделать следующие выводы:
1. Средняя ЗП сотрудников составляет 137570 USD, минимальная 5132 USD, максимальная 450000 USD.Половина сотрудникков имею ЗП в диапазоне от 95000 USD до 175000 USD.
2. Более 50 % (1923) сотрудников компании работают только удаленно.Около 5% (189) сотрудников работают и удаленно и в офисе.Остальные сотрудники (1923) работают только в офисе.
3.